internal/ansible/metrics/metrics.go

```diff
@@ -82,6 +85,162 @@ func RegisterBuildInfo(r prometheus.Registerer) {
+}
+
+func handleGauge(metricSpec UserMetric, gauge prometheus.Gauge) error {
+	if metricSpec.Gauge == nil {
+		return fmt.Errorf("cannot change metric type of %s, which is a gauge", metricSpec.Name)
+	}
+	if metricSpec.Gauge.Inc {
+		gauge.Inc()
+	} else if metricSpec.Gauge.Dec {
+		gauge.Dec()
+	} else if metricSpec.Gauge.Add != 0.0 {
+		gauge.Add(metricSpec.Gauge.Add)
+	} else if metricSpec.Gauge.Sub != 0.0 {
+		gauge.Sub(metricSpec.Gauge.Sub)
+	} else if metricSpec.Gauge.Set != 0.0 {
+		gauge.Set(metricSpec.Gauge.Set)
+	} else if metricSpec.Gauge.SetToCurrentTime {
+		gauge.SetToCurrentTime()
+	}
+	return nil
+}
+
+func handleSummaryOrHistogram(metricSpec UserMetric, summary prometheus.Summary) error {
+	if metricSpec.Histogram == nil && metricSpec.Summary == nil {
+		return fmt.Errorf("cannot change metric type of %s, which is a histogram or summary", metricSpec.Name)
+	}
+	if metricSpec.Histogram != nil {
+		summary.Observe(metricSpec.Histogram.Observe)
+	} else if metricSpec.Summary != nil {
+		summary.Observe(metricSpec.Summary.Observe)
+	}
+	return nil
+}
+
+func ensureMetric(r prometheus.Registerer, metricSpec UserMetric) {
+	if _, ok := userMetrics[metricSpec.Name]; !ok {
+		// This is the first time we've seen this metric
+		logf.Log.WithName("metrics").Info("Registering", "metric", metricSpec.Name)
+		if metricSpec.Counter != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewCounter(prometheus.CounterOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Gauge != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewGauge(prometheus.GaugeOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Histogram != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewHistogram(prometheus.HistogramOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Summary != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewSummary(prometheus.SummaryOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if err := r.Register(userMetrics[metricSpec.Name]); err != nil {
+			logf.Log.WithName("metrics").Info("Unable to register %s metric with prometheus.", metricSpec.Name)
+		}
+	}
+}
+
+func HandleUserMetric(r prometheus.Registerer, metricSpec UserMetric) error {
+	if err := validateMetricSpec(metricSpec); err != nil {
+		return err
+	}
+	ensureMetric(r, metricSpec)
+	collector := userMetrics[metricSpec.Name]
+	switch v := collector.(type) {
+	// Gauge must be first, because a Counter is a Gauge, but a Gauge is not a Counter.
+	case prometheus.Gauge:
+		if err := handleGauge(metricSpec, v); err != nil {
+			return err
+		}
+	case prometheus.Counter:
+		if err := handleCounter(metricSpec, v); err != nil {
+			return err
+		}
+	// Histogram and Summary interfaces are identical, so we accept either case.
+	case prometheus.Histogram:
+		if err := handleSummaryOrHistogram(metricSpec, v); err != nil {
+			return err
+		}
+	}
```

Comment:
Instead of the type switch, you can do:
```
switch {
case metricSpec.Gauge != nil:
    // gauge stuff
case metricSpec.Counter != nil:
    // counter stuff
// etc
}
```
Not a blocker for this PR, but would allow you to split the summary and histogram helper functions.

Reply:
I dont think that will work in edge cases where the user is erroneously mixing types. For example, if the user creates a metric "my_counter_metric", and then calls the gauge function on "my_counter_metric". In this case, the current code will see that it is a counter, and handle it as a counter, and send a failure message that "you cant change counter to gauge". For your suggestion, it will attempt to handle it as a gauge, which could cause silent problems I think?

Reply:
Gotcha, do we validate anywhere else that for existing metrics, the type doesn't change?

Reply:
We didn't collect custom metrics from users, so AFAIK it hasn't come up before.